cranelift-codegen/src/isa/x86/abi.rs

```diff
@@ -752,6 +820,36 @@ fn insert_common_prologue(
         }
     }
 
+    for reg in csrs.iter(FPR) {
+        // Append param to entry EBB
+        let csr_arg = pos.func.dfg.append_ebb_param(ebb, types::F64X2);
+
+        // Assign it a location
+        pos.func.locations[csr_arg] = ir::ValueLoc::Reg(reg);
+
+        let reg_store_inst = pos.ins().store(
+            ir::MemFlags::trusted(),
+            csr_arg,
+            panic!("get RSP value"),
```

Comment:
this is where I'd like input from someone who knows codegen APIs better than me (@sstangl?): SIMD registers can't be `x86_push`'d nor `x86_pop`'d (there is no push/pop encoding in the ISA for them, even), so my plan is to know there will be sufficient space reserved for them and simply `load`/`store` them. To do that, I would need to use `rsp` or `rbp` as a base, with an appropriate offset. Is there a nice way to get `function call frame base` as a value?

I don't want to load `rbp` here (even if it were possible) because future frame pointer elision would break this, so either `rsp` or `function call frame base` seem like better choices. Computing the offset from here ought to be straightforward.

Reply:
I think [the IonMonkey-style approach](https://searchfox.org/mozilla-central/rev/cbd110d718bc89a499d3f996af24532abbf6f8ea/js/src/jit/x64/Trampoline-x64.cpp#72) works well: figure out how many registers to push, allocate stack by subtracting from `%rsp`, and `vmovdqa` into place in an arbitrary order of your choosing.

Reply:
Yes, I think that would work well here too. What I don't know is how to get the value of `rsp` as a value I can provide here! Is there a way to write "give me an SSA value that is whatever is in `rsp` at this point"? The other option I can imagine is another pair of `x86_*` special instructions that directly encode `rsp`, and take the offset as an argument.

Sidenote: I see IonMonkey uses `vmovdqa` in favor of `movdqa` - is that actually what gets emitted? Is there a reason IonMonkey prefers the AVX form?

Reply:
Would it be possible to use `stack_store`?

Reply:
ah! yes, I think so. If not, `stack_addr` is the piece I was missing before, and should work too.

Reply:
it turns out `stack_load` and `stack_store` won't do - there's no direct encoding for them in x86 (yet, I hope) and `prologue_epilogue` is well past `legalize`.

`stack_addr` was surprising too: we're also past `regalloc` so I have to assign a location myself. The error not doing this (that I violated constraints) was quite confusing, because .. what constraints would copying `rsp` have? :D 

A third consideration I didn't realize: we're also well past `postopt`, so opportunity to merge a `stack_addr/{load,store}` has passed. To this end, I added only one `stack_addr` in the prologue, and one in the epilogue, and only if we actually preserve any FPRs in the first place. With `stack_load`/`stack_store` encodings this can be cleaned up some more, but this already took longer than I'd anticipated.

Anyway, it works now!